### PR TITLE
Add more typings:

### PR DESCRIPTION
**Pull request by bryanforbes**
* sql.elements
* sql.functions
* sql.selectable
* sql.sqltypes
* sql.type_api

### CODE REVIEW

**Comment by ilevkivskyi** [comment]
Sorry for a delay. I probably will not have time to review this today, but will try to do this tomorrow.

File: sqlalchemy-stubs/sql/elements.pyi [hunk 1]
```diff
@@ -334,34 +334,47 @@ _G = TypeVar('_G', bound=Grouping)
 
 class Grouping(ColumnElement[_T]):
     __visit_name__: str = ...
-    element: ColumnElement[_T] = ...
+    element: ClauseElement = ...
     type: TypeEngine[_T] = ...
+    @overload
     def __init__(self, element: ColumnElement[_T]) -> None: ...
+    @overload
+    def __init__(self, element: ClauseElement) -> None: ...
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/sql/elements.pyi]: At runtime, what will be `_T` if `ClauseElement` is given to `__init__`?

File: sqlalchemy-stubs/sql/functions.pyi [hunk 2]
```diff
@@ -15,49 +16,60 @@ def register_function(identifier, fn, package: str = ...): ...
...
+    @overload
+    def filter(self) -> FunctionElement[_T]: ...
+    @overload
+    def filter(self, criteria: Any, *criterion: Any) -> FunctionFilter[_T]: ...
+    def get_children(self, **kwargs: Any) -> Tuple[Grouping[Any]]: ...
+    def within_group_type(self, within_group: Any) -> Any: ...
+    def alias(self, name: Optional[str] = ..., flat: bool = ...) -> Alias: ...
+    def select(self) -> Select: ...  # type: ignore  # incompatible with FromClause.select
+    def scalar(self) -> Any: ...  # type: ignore  # incompatible with Executable.scalar
+    def execute(self) -> ResultProxy: ...  # type: ignore  # incompatible with Executable.scalar
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/sql/functions.pyi]: Typo: should be `Executable.execute` here?

File: sqlalchemy-stubs/sql/functions.pyi [hunk 3]
```diff
@@ -15,49 +16,60 @@ def register_function(identifier, fn, package: str = ...): ...
...
+    def self_group(self, against: Optional[Any] = ...) -> Union[ColumnElement[_T], Grouping[_T]]: ...
 
 class _FunctionGenerator(object):
     opts: Any = ...
-    def __init__(self, **opts) -> None: ...
-    def __getattr__(self, name): ...
-    def __call__(self, *c, **kwargs): ...
+    def __init__(self, **opts: Any) -> None: ...
+    def __getattr__(self, name: str) -> Any: ...
+    def __call__(self, *c: Any, **kwargs: Any) -> Function[_T]: ...
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/sql/functions.pyi]: What is `_T` here? This class is not generic currently.

File: sqlalchemy-stubs/sql/sqltypes.pyi [hunk 4]
```diff
@@ -77,47 +85,52 @@ class Float(Numeric):
...
+    def python_type(self) -> Type[time]: ...
 
-class _Binary(TypeEngine[bytes]):
-    length: Any = ...
-    def __init__(self, length: Optional[Any] = ...) -> None: ...
-    def literal_processor(self, dialect): ...
+if sys.version_info[0] < 3:
+    _BinaryType = str
+else:
+    _BinaryType = bytes
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/sql/sqltypes.pyi]: Is this really needed? I though `bytes` is always the binary type, in Python 2 it is just an alias for `str`

File: sqlalchemy-stubs/sql/sqltypes.pyi [hunk 5]
```diff
@@ -203,26 +216,30 @@ class JSON(Indexable, TypeEngine[Dict[str, Any]]):
     class JSONPathType(JSONElementType): ...
     comparator_factory: Any = ...
     @property
-    def python_type(self): ...
+    def python_type(self) -> Type[dict]: ...
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/sql/sqltypes.pyi]: I think it should be `Dict[str, Any]`.

File: sqlalchemy-stubs/sql/sqltypes.pyi [hunk 6]
```diff
@@ -203,26 +216,30 @@ class JSON(Indexable, TypeEngine[Dict[str, Any]]):
...
     @property
-    def should_evaluate_none(self): ...
+    def should_evaluate_none(self) -> bool: ...  # type: ignore  # incompatible with supertype
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/sql/sqltypes.pyi]: It is always better to mention which supertype it is incompatible with (here and below).

File: sqlalchemy-stubs/sql/functions.pyi (hunk 3 above)
Comment (bryanforbes) [reply sqlalchemy-stubs/sql/functions.pyi]: This should probably be changed to `Any`

File: sqlalchemy-stubs/sql/sqltypes.pyi (hunk 4 above)
Comment (bryanforbes) [reply sqlalchemy-stubs/sql/sqltypes.pyi]: I was just following what sqlalchemy does in `sqlalchemy.util.compat`. If you think this should just be bytes, I can change it to that.

**Comment by bryanforbes** [comment]
> Also, I finally got time to run this against our internal codebase. This gives six extra errors, among those:
> 
>     * `Argument "convert_unicode" to "UnicodeText" has incompatible type "str"; expected "bool"`
> 
>     * `Argument "convert_unicode" to "Unicode" has incompatible type "str"; expected "bool"`

I've fixed the signature of `__init__` for these types and have added tests for each of them locally.

>     * `"Alias" has no attribute "__iter__" (not iterable)`

I couldn't find an `__iter__` in the class hierarchy of `Alias`, so I tried the following in a REPL:

```python
from sqlalchemy import MetaData, Table, Column, Integer

m = MetaData()
t = Table('foo', m, Column('id', Integer))
a = t.alias('bar')
for o in a:    # Type Error: 'Alias' object is not iterable
    print(o)
```

Is your codebase adding something to `Alias`?

> Are any tests needed for this PR, what do you think?

I've been adding tests for the failures you've found in the codebase you're checking. I can add more if you think we need more.

File: sqlalchemy-stubs/sql/sqltypes.pyi (hunk 4 above)
Comment (ilevkivskyi) [reply sqlalchemy-stubs/sql/sqltypes.pyi]: OK, let us keep it then.

File: sqlalchemy-stubs/sql/elements.pyi (hunk 1 above)
Comment (ilevkivskyi) [reply sqlalchemy-stubs/sql/elements.pyi]: Could you please a comment about this to the code?

**Comment by ilevkivskyi** [comment]
> Is your codebase adding something to `Alias`?

Maybe, I didn't look in detail, so let us ignore this for now.

> I can add more if you think we need more.

No, I think just a few about the `__init__` is enough, thanks.

> I've fixed the signature of `__init__` for these types and have added tests for each of them locally.

You can push the latest updates, hopefully we can merge this today.

File: sqlalchemy-stubs/sql/elements.pyi (hunk 1 above)
Comment (bryanforbes) [reply sqlalchemy-stubs/sql/elements.pyi]: I practice, I don't believe that `Grouping()` is directly used (it's not exposed from the main module) and is only exposed from the `self_group()` method of some subclasses of `ClauseElement`. So in the general case, those methods are going to have `Grouping` typed in their return type. For the unusual case where `Grouping()` is instantiated by the user with a ClauseElement, an error will be generated by mypy:

```python
from sqlalchemy.sql.elements import Grouping, TextClause

g = Grouping(TextClause('select * from users'))  # error: Need type annotation for 'g'
```

Comment (bryanforbes) [reply sqlalchemy-stubs/sql/elements.pyi]: I have added a comment above the signature:

```python
    # the following __init__ signature produces Grouping[None] at runtime but mypy will require an annotation
    @overload
    def __init__(self, element: ClauseElement) -> None: ...
```

**Pull request merged**